corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js

```diff
@@ -233,9 +242,147 @@ hqDefine("geospatial/js/case_grouping_map",[
+            const groupId = caseItem.groupId;
+            if (groupId) {
+                if (groupColors[groupId] === undefined) {
+                    groupColors[groupId] = getRandomColor();
+                }
+                const color = groupColors[groupId];
+
+                const marker = new mapboxgl.Marker({ color: color, draggable: false });  // eslint-disable-line no-undef
+                marker.setLngLat([caseItem.coordinates.lng, caseItem.coordinates.lat]);
+
+                // Add the marker to the map
+                marker.addTo(map);
+                mapMarkers.push(marker);
+            }
+        });
+    }
+
+    function uuidv4() {
+        // https://stackoverflow.com/questions/105034/how-do-i-create-a-guid-uuid/2117523#2117523
+        return "10000000-1000-4000-8000-100000000000".replace(/[018]/g, c =>
+            (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16),
+        );
+    }
+
+    async function setCaseGroups() {
+        const sourceFeatures = map.querySourceFeatures('caseWithGPS', {
+            sourceLayer: 'clusters',
+            filter: ['==', 'cluster', true],
+        });
+        const clusterSource = map.getSource('caseWithGPS');
+        let caseGroups = {};
+
+        for (const cluster of sourceFeatures) {
+            const clusterId = cluster.properties.cluster_id;
+            const pointCount = cluster.properties.point_count;
+
+            try {
+                const casePoints = await getClusterLeavesAsync(clusterSource, clusterId, pointCount);
+                const groupUUID = uuidv4();
+                for (const casePoint of casePoints) {
+                    const caseId = casePoint.properties.id;
+                    caseGroups[caseId] = {
+                        groupId: groupUUID,
+                        groupCoordinates: {
+                            lng: cluster.geometry.coordinates[0],
+                            lat: cluster.geometry.coordinates[1],
+                        },
+                    };
+                }
+            } catch (error) {
+                console.error("Error processing cluster:", error);
+            }
+        }
+        loadCaseGroupsInExport(caseGroups);
+        collapseGroupsOnMap();
+    }
+
+    function clearCaseGroups() {
+        setMapLayersVisibility(MAPBOX_LAYER_VISIBILITY.Visible);
+        mapMarkers.forEach((marker) => marker.remove());
+        mapMarkers = [];
+        clearCaseGroupsInExport();
+    }
+
+    function groupLockModel() {
+        'use strict';
+        var self = {};
+
+        self.groupsLocked = ko.observable(false);
+
+        self.showLockButton = ko.computed(function () {
+            return !self.groupsLocked();
+        });
+        self.showUnlockButton = ko.computed(function () {
+            return self.groupsLocked();
+        });
+
+        self.toggleGroupLock = function () {
+            if (self.groupsLocked()) {
+                self.groupsLocked(false);
```

Comment:
Nit: We could remove some code duplication here and simply put this at the top of the function as `self.groupsLocked(!self.groupsLocked())`.

Reply:
Yes!

Reply:
Hehe, such a rookie mistake